server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java

```diff
@@ -104,54 +115,78 @@ public void handle(Request request, Response response) throws Exception {
+      .setComponent(request.param(PARAM_COMPONENT));
+    if (request.hasParam(PARAM_KEYS)) {
+      builder.setKeys(request.paramAsStrings(PARAM_KEYS));
+    }
+    return builder.build();
   }
 
-  private Optional<ComponentDto> getComponent(DbSession dbSession, ValuesRequest valuesRequest) {
-    String componentId = valuesRequest.getComponentId();
-    String componentKey = valuesRequest.getComponentKey();
-    if (componentId != null || componentKey != null) {
-      return Optional.of(componentFinder.getByUuidOrKey(dbSession, componentId, componentKey, ID_AND_KEY));
+  private Set<String> loadKeys(ValuesRequest valuesRequest) {
+    List<String> keys = valuesRequest.getKeys();
+    if (keys.isEmpty()) {
+      return concat(
+        concat(loadLicenseHashKeys(), propertyDefinitions.getAll().stream().map(PropertyDefinition::key)),
+        ADDITIONAL_KEYS.stream())
+          .collect(Collectors.toSet());
     }
-    return Optional.empty();
+    return new HashSet<>(keys);
   }
 
-  private void checkAdminPermission(Optional<ComponentDto> component) {
-    if (component.isPresent()) {
-      userSession.checkComponentUuidPermission(UserRole.ADMIN, component.get().uuid());
-    } else {
-      userSession.checkPermission(GlobalPermissions.SYSTEM_ADMIN);
+  private Stream<String> loadLicenseHashKeys() {
+    return propertyDefinitions.getAll()
+      .stream()
+      .filter(setting -> setting.type().equals(LICENSE))
+      .map(PropertyDefinition::key)
+      .map(setting -> setting.replace(LICENSE_SUFFIX, LICENSE_HASH_SUFFIX));
+  }
+
+  private Optional<ComponentDto> loadComponent(DbSession dbSession, ValuesRequest valuesRequest) {
+    String componentKey = valuesRequest.getComponent();
+    if (componentKey != null) {
+      ComponentDto component = componentFinder.getByKey(dbSession, componentKey);
+      userSession.checkComponentUuidPermission(USER, component.projectUuid());
+      return Optional.of(component);
     }
+    return Optional.empty();
   }
 
   private List<Setting> loadSettings(DbSession dbSession, Optional<ComponentDto> component, Set<String> keys) {
     // List of settings must be kept in the following orders : default -> global -> component
     List<Setting> settings = new ArrayList<>();
     settings.addAll(loadDefaultSettings(keys));
     settings.addAll(settingsFinder.loadGlobalSettings(dbSession, keys));
-    if (component.isPresent()) {
-      settings.addAll(settingsFinder.loadComponentSettings(dbSession, keys, component.get()).values());
-    }
-    return settings;
+    component.ifPresent(componentDto -> settings.addAll(settingsFinder.loadComponentSettings(dbSession, keys, componentDto).values()));
+    return settings.stream()
+      .filter(isVisible(component))
+      .collect(Collectors.toList());
+  }
+
+  private Predicate<Setting> isVisible(Optional<ComponentDto> component) {
+    return setting -> !setting.getKey().endsWith(SECURED_SUFFIX)
```

Comment:
remark: don't you have the feeling our setting objects should be smarter? We should be able to have setting.isSecured() and encapsulate the logic
I know we use Protobuf objects, but we should try to find a better design if it's not directly with setting.isSecured()